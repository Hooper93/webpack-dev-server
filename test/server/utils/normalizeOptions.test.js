'use strict';

const webpack = require('webpack');
const normalizeOptions = require('../../../lib/utils/normalizeOptions');

describe('normalizeOptions', () => {
  const cases = [
    {
      title: 'no options',
      multiCompiler: false,
      options: {},
      optionsResults: null,
    },
    {
      title: 'contentBase string',
      multiCompiler: false,
      options: {
        contentBase: '/path/to/dist',
      },
      optionsResults: null,
    },
    {
      title: 'contentBase array',
      multiCompiler: false,
      options: {
        contentBase: ['/path/to/dist1', '/path/to/dist2'],
      },
      optionsResults: null,
    },
    {
      title: 'watchOptions',
      multiCompiler: false,
      options: {
        watchOptions: {
          poll: true,
        },
      },
      optionsResults: null,
    },
    {
      title: 'transportMode sockjs string',
      multiCompiler: false,
      options: {
        transportMode: 'sockjs',
      },
      optionsResults: null,
    },
    {
      title: 'transportMode ws string',
      multiCompiler: false,
      options: {
        transportMode: 'ws',
      },
      optionsResults: null,
    },
    {
      title: 'transportMode ws object',
      multiCompiler: false,
      options: {
        transportMode: {
          server: 'ws',
          client: 'ws',
        },
      },
      optionsResults: null,
    },
    {
      title: 'transportMode custom server path',
      multiCompiler: false,
      options: {
        transportMode: {
          server: '/path/to/custom/server/',
        },
      },
      optionsResults: null,
    },
    {
      title: 'transportMode custom server class',
      multiCompiler: false,
      options: {
        transportMode: {
          server: class CustomServerImplementation {},
        },
      },
      optionsResults: null,
    },
    {
      title: 'transportMode custom client path',
      multiCompiler: false,
      options: {
        transportMode: {
          client: '/path/to/custom/client/',
        },
      },
      optionsResults: null,
    },
    {
      title: 'client host and port',
      multiCompiler: false,
      options: {
        client: {
          host: 'my.host',
          port: 9000,
        },
      },
      optionsResults: null,
    },
    {
      title: 'client path',
      multiCompiler: false,
      options: {
        client: {
          path: '/custom/path/',
        },
      },
      optionsResults: null,
    },
    {
      title: 'client path without leading/ending slashes',
      multiCompiler: false,
      options: {
        client: {
          path: 'custom/path',
        },
      },
      optionsResults: null,
    },
    {
      title: 'liveReload is true',
      multiCompiler: false,
      options: {
        liveReload: true,
      },
      optionsResults: null,
    },
    {
      title: 'liveReload is false',
      multiCompiler: false,
      options: {
        liveReload: false,
      },
      optionsResults: null,
    },
    {
      title: 'hot is true',
      multiCompiler: false,
      options: {
        hot: true,
      },
      optionsResults: null,
    },
    {
      title: 'hot is false',
      multiCompiler: false,
      options: {
        hot: false,
      },
      optionsResults: null,
    },
    {
      title: 'hot is only',
      multiCompiler: false,
      options: {
        hot: 'only',
      },
      optionsResults: null,
    },
    {
      title: 'dev is set',
      multiCompiler: false,
      options: {
        dev: {
          serverSideRender: true,
        },
      },
      optionsResults: null,
    },
    {
      title: 'static is true',
      multiCompiler: false,
      options: {
        static: true,
      },
      optionsResults: null,
    },
    {
      title: 'static is false',
      multiCompiler: false,
      options: {
        static: false,
      },
      optionsResults: null,
    },
    {
      title: 'static is string',
      multiCompiler: false,
      options: {
        static: '/static/path',
      },
      optionsResults: null,
    },
    {
      title: 'static is an array of strings',
      multiCompiler: false,
      options: {
        static: ['/static/path1', '/static/path2'],
      },
      optionsResults: null,
    },
    {
      title: 'static is an array of static objects',
      multiCompiler: false,
      options: {
        static: [
          {
            directory: '/static/path1',
          },
          {
            publicPath: '/static/public/path',
          },
        ],
      },
      optionsResults: null,
    },
    {
      title: 'static is an array of strings and static objects',
      multiCompiler: false,
      options: {
        static: [
          '/static/path1',
          {
            publicPath: '/static/public/path',
          },
        ],
      },
      optionsResults: null,
    },
  ];

  cases.forEach((data) => {
    describe(data.title, () => {
      let compiler;
      beforeAll(() => {
        let webpackConfig;
        if (data.multiCompiler) {
          webpackConfig = require('../../fixtures/multi-compiler-config/webpack.config');
        } else {
          webpackConfig = require('../../fixtures/simple-config/webpack.config');
        }

        compiler = webpack(webpackConfig);
      });

      it('should set correct options', () => {
        const originalContentBase = data.options.contentBase;
        normalizeOptions(compiler, data.options);
        if (data.optionsResults) {
          expect(data.options).toEqual(data.optionsResults);
        } else {
          if (data.options.contentBase !== originalContentBase) {
            // we don't want this in the snapshot, because it is
            // the current working directory
            expect(data.options.contentBase).toEqual(process.cwd());
            delete data.options.contentBase;
          }

          if (data.options.static) {
            data.options.static.forEach((staticOpts) => {
              if (staticOpts.directory === process.cwd()) {
                // give an indication in the snapshot that this is the
                // current working directory
                staticOpts.directory = 'CWD';
              }
            });
          }

          expect(data.options).toMatchSnapshot();
        }
      });
    });
  });
});
