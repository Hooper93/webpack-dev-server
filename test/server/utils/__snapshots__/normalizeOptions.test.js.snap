// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeOptions client host and port should set correct options 1`] = `
Object {
  "client": Object {
    "host": "my.host",
    "path": "/ws",
    "port": 9000,
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions client path should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/custom/path",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions client path without leading/ending slashes should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/custom/path",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions contentBase array should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBase": Array [
    "/path/to/dist1",
    "/path/to/dist2",
  ],
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions contentBase string should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBase": "/path/to/dist",
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions contentBasePublicPath string should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/content-base-public-path",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions hot is false should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": false,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions hot is only should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": "only",
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions hot is true should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions liveReload is false should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": false,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions liveReload is true should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions no options should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode custom client path should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "/path/to/custom/client/",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode custom server class should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": [Function],
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode custom server path should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "/path/to/custom/server/",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode sockjs string should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "sockjs",
    "server": "sockjs",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode ws object should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions transportMode ws string should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {},
}
`;

exports[`normalizeOptions watchOptions should set correct options 1`] = `
Object {
  "client": Object {
    "path": "/ws",
  },
  "contentBasePublicPath": "/",
  "hot": true,
  "liveReload": true,
  "serveIndex": true,
  "stats": Object {},
  "transportMode": Object {
    "client": "ws",
    "server": "ws",
  },
  "watchOptions": Object {
    "poll": true,
  },
}
`;
